### PR TITLE
Add new scripting language for the config

### PR DESCRIPTION
**Pull request by TimUntersberger**
Since this is done soon, I am now opening a PR.

This will be a massive PR so I don't expect @ramirezmike to look through this tbh.

I still have to finish the documentation and fix 1 bug then I'm done.

TODO:

- [x] replace placeholder version variable
- [x] Finish documentation
- [x] Fix custom modes
- [x] Fix workflows
- [x] add a workflow that automatically builds the source files of the doc-gen folder and moves the output to its place in the book folder and then builds the book (Maybe done?)
- [x] Implement bind_map
- [x] Implement bind_arr
- [x] fix review problems

### CODE REVIEW

File: doc-gen/output/api/nog/toggle_mode.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# toggle_mode
+
+Toggles the work mode.
```
Comment (ramirezmike) [inline doc-gen/output/api/nog/toggle_mode.md:3]: this should be "Toggle the mode"

**Comment by ramirezmike** [comment]
Not sure where the best place for this is, whether you want separate issues or one like... regression issue? But, I wrote down things I ran into while using latest today.

1) Alt+0 binding for change workspace
    I have **nog.bind_arr("Alt", nog.workspace.change, range(workspace_count))** in my config based kinda on what I see in the docs and your example and when I do that Alt+0 doesn't move to the 10th workspace, it just doesn't do anything. I think it's because range returns 0-9 and then with the bind_arr it binds 0 to workspace 0 which doesn't exist. I don't know if this is really a bug, it's just like... maybe can be remedied with documentation changes. The "workaround" is to just bind Alt+0 to change to workspace 10 and then use the range afterward for everything else. Not sure if you ran into this?

2) nog.launch is hit or miss for me
I haven't looked into this much just noticed that some of the same strings I had in my old config no longer work but it still works for some things like just "notepad.exe"

3) nog.config.toggle
This just doesn't work for me. I've been digging into this one a bit today because I like having this for toggling visibility of the appbar. I realized that the toggle code in src/nogscript/lib/mod.rs is "toggling" from an in memory version of the initially loaded config 

                let mut cfg = cfg.lock().clone();
                vs
                let mut cfg = state.lock().config.clone();

I noticed this in a couple places but after changing that the appbar started toggling the way I expected _BUT_ I started running into issues with the hot reload code, the unregistering/registering keybindings and the state locks which I'm looking at a bit.

4) modifying config in mode
I think this is similar to #168 

5) min_width/min_height
I haven't dug into the code here but I noticed apps weren't respecting my settings for this. Might look at it more tomorrow in case it's just a config issue

6) ignore_fullscreen_actions
While working on the move_in/move_out I saw some code was missing the checks for this. I haven't tested it out but I suspect there might be some places where this check needs to be re-added 

7) ignore rules not working?
I can't get the ignore to work following the example config

8) move to monitor crashes
This I noticed while looking at the AltGr thing and then noticed again looking at issue 1 in this list. If you accidentally move to a monitor that doesn't exist (who doesn't have 10 monitors though, right?) nog crashes

Hope that doesn't sound too doom & gloom, I actually am digging the config changes overall.

**Comment by TimUntersberger** [comment]
> nog.launch is hit or miss for me

This should be techincally the same. Can you give me some examples of things that are not working?

**Comment by ramirezmike** [comment]
for ignore and launch, I messed with it some more and it works for more programs than it did before. I think some of my keybindings were being grabbed by other apps, but I'll check again later. These are probably OK.

>I changed the thing you mentioned and I don't encounter the issue you mention. Can you give me an example?

I see you updated so I pulled latest from development. I have this in my config.

    nog.config.enable("display_app_bar");
    nog.bind("Alt+I", () => nog.config.toggle("display_app_bar"))

So, that fixed the issue where it was not actually toggling but what I found is that if you time it right nog ends up in this deadlock state and I have to kill it with task manager. It's like the hot_reload code is locking the state/config while the appbar is refreshing. The deadlock doesn't occur at all if I comment out the refresh thread at the bottom of ~/twm/src/window.rs. It's tricky because... I'd assume that state.lock() should just block the thread until it's ready to own it, but it's like if you time it right, the appbar never finishes refreshing and lets go of its state lock.

It doesn't happen every time but if you try toggling a bit you'll likely hit it. The same thing happens when entering/leaving work mode.

**Comment by TimUntersberger** [comment]
> It doesn't happen every time but if you try toggling a bit you'll likely hit it. The same thing happens when entering/leaving work mode.

I just spammed the `F1` binding from above for about a minute and never hit it. Idk how to reproduce this.

**Pull request merged**